interfaces/hotplug/deviceinfo.go

```
@@ -0,0 +1,113 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package hotplug
+
+import (
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+)
+
+type HotplugDeviceInfo struct {
+	object       string
+	Data         map[string]string
+	idVendor     string
+	idProduct    string
+	product      string
+	manufacturer string
+	serial       string
+}
+
+func NewHotplugDeviceInfo(obj string, env map[string]string) *HotplugDeviceInfo {
+	return &HotplugDeviceInfo{
+		object: obj,
+		Data:   env,
+	}
+}
+
+func (h *HotplugDeviceInfo) Object() string {
+	return h.object
+}
+
+func (h *HotplugDeviceInfo) Path() string {
+	return filepath.Join(dirs.SysDir, h.Data["DEVPATH"])
+}
+
+func (h *HotplugDeviceInfo) Subsystem() string {
+	return h.Data["SUBSYSTEM"]
+}
+
+func (h *HotplugDeviceInfo) Minor() string {
+	return h.Data["MINOR"]
+}
+
+func (h *HotplugDeviceInfo) Major() string {
+	return h.Data["MAJOR"]
+}
+
+func (h *HotplugDeviceInfo) DeviceName() string {
+	return h.Data["DEVNAME"]
+}
+
+func (h *HotplugDeviceInfo) DeviceType() string {
+	return h.Data["DEVTYPE"]
+}
+
+func (h *HotplugDeviceInfo) IdVendor() string {
+	return h.readOnceMaybe("idVendor", &h.idVendor)
+}
+
+func (h *HotplugDeviceInfo) IdProduct() string {
+	return h.readOnceMaybe("idProduct", &h.idProduct)
+}
+
+func (h *HotplugDeviceInfo) Product() string {
+	return h.readOnceMaybe("product", &h.product)
+}
+
+func (h *HotplugDeviceInfo) Manufacturer() string {
+	return h.readOnceMaybe("manufacturer", &h.manufacturer)
+}
+
+func (h *HotplugDeviceInfo) Serial() string {
+	return h.readOnceMaybe("serial", &h.serial)
+}
+
+func (h *HotplugDeviceInfo) readOnceMaybe(fileName string, out *string) string {
+	if *out == "" {
+		data, err := ioutil.ReadFile(filepath.Join(h.Path(), fileName))
+		if err != nil {
```

I'm unsure if we should just ignore all errors this way. Sysfs has some very special semantics on how to traverse and read stuff from it correctly. Are you familiar with https://www.kernel.org/doc/html/v4.16/admin-guide/sysfs-rules.html

---

Removed for now per our discussion and the comment above.